File: pc/voicing_protocol.py
DEFAULT_SERVER_IP = "192.168.137.1"
WEBSOCKET_PORT = 9527
UDP_BROADCAST_PORT = 9530

UDP_DISCOVERY_TYPE = "voice_coding_server"

TYPE_TEXT = "text"
TYPE_PING = "ping"
TYPE_CONNECTED = "connected"
TYPE_ACK = "ack"
TYPE_PONG = "pong"
TYPE_SYNC_STATE = "sync_state"
TYPE_SYNC_DISABLED = "sync_disabled"
QR_SCAN_PING_SOURCE = "qr_scan"
QR_PAYLOAD_TYPE = "voicing"
QR_PAYLOAD_VERSION = 1
TEXT_SEND_MODE_SUBMIT = "submit"
TEXT_SEND_MODE_SHADOW = "shadow"
TEXT_SEND_MODE_COMMIT = "commit"

CLIENT_TO_SERVER_TYPES = {
    TYPE_TEXT,
    TYPE_PING,
}

SERVER_TO_CLIENT_TYPES = {
    TYPE_CONNECTED,
    TYPE_ACK,
    TYPE_PONG,
    TYPE_SYNC_STATE,
    TYPE_SYNC_DISABLED,
}


def build_qr_payload(
    *,
    device_id: str,
    ip: str,
    port: int,
    name: str,
    os_name: str,
    ips: list[str] | None = None,
) -> dict:
    payload = {
        "v": QR_PAYLOAD_VERSION,
        "type": QR_PAYLOAD_TYPE,
        "device_id": device_id,
        "ip": ip,
        "port": port,
        "name": name,
        "os": os_name,
    }
    if ips:
        payload["ips"] = ips
    return payload


def build_connected_message(
    sync_enabled: bool,
    computer_name: str,
    device_id: str = "",
    os_name: str = "",
) -> dict:
    return {
        "type": TYPE_CONNECTED,
        "message": "Connected to Voicing server",
        "sync_enabled": sync_enabled,
        "computer_name": computer_name,
        "device_id": device_id,
        "name": computer_name,
        "os": os_name,
    }


def build_ack_message(clear_input: bool = True) -> dict:
    return {
        "type": TYPE_ACK,
        "message": "Text received and typed",
        "clear_input": clear_input,
    }


def build_pong_message(sync_enabled: bool) -> dict:
    return {
        "type": TYPE_PONG,
        "sync_enabled": sync_enabled,
    }


def build_sync_state_message(sync_enabled: bool) -> dict:
    return {
        "type": TYPE_SYNC_STATE,
        "sync_enabled": sync_enabled,
    }


def build_sync_disabled_message() -> dict:
    return {
        "type": TYPE_SYNC_DISABLED,
        "sync_enabled": False,
        "message": "Sync is disabled on PC",
    }


def build_text_message(
    content: str,
    auto_enter: bool = False,
    send_mode: str = TEXT_SEND_MODE_SUBMIT,
) -> dict:
    return {
        "type": TYPE_TEXT,
        "content": content,
        "auto_enter": auto_enter,
        "send_mode": send_mode,
    }


def build_ping_message() -> dict:
    return {
        "type": TYPE_PING,
    }

File: android/voice_coding/lib/voicing_protocol.dart
import 'dart:convert';

class VoicingProtocol {
  static const String defaultServerIp = '192.168.137.1';
  static const int websocketPort = 9527;
  static const int udpBroadcastPort = 9530;
  static const int bluetoothRfcommChannel = 11;
  static const int heartbeatIntervalSec = 15;
  static const int heartbeatTimeoutSec = 30;
  static const int connectTimeoutSec = 8;
  static const int maxReconnectDelaySec = 30;
  static const int udpReconnectCooldownMs = 3000;
  static const String bluetoothServiceUuid =
      '8b3e3f4b-6f8f-4f2f-9d5d-77f4f84f9d11';
  static const String bluetoothMessageDelimiter = '\n';

  static const String udpDiscoveryType = 'voice_coding_server';

  static const String typeText = 'text';
  static const String typePing = 'ping';
  static const String typeConnected = 'connected';
  static const String typeAck = 'ack';
  static const String typePong = 'pong';
  static const String typeSyncState = 'sync_state';
  static const String typeSyncDisabled = 'sync_disabled';
  static const String textSendModeSubmit = 'submit';
  static const String textSendModeShadow = 'shadow';
  static const String textSendModeCommit = 'commit';

  static const Set<String> clientMessageTypes = {
    typeText,
    typePing,
  };

  static const Set<String> serverMessageTypes = {
    typeConnected,
    typeAck,
    typePong,
    typeSyncState,
    typeSyncDisabled,
  };

  static Map<String, dynamic> buildTextMessage(
    String content, {
    bool autoEnter = false,
    String sendMode = textSendModeSubmit,
  }) {
    return {
      'type': typeText,
      'content': content,
      'auto_enter': autoEnter,
      'send_mode': sendMode,
    };
  }

  static Map<String, dynamic> buildPingMessage() {
    return {'type': typePing};
  }

  static Map<String, dynamic>? decodeMessage(dynamic message) {
    if (message is! String) {
      return null;
    }

    final decoded = json.decode(message);
    return decoded is Map<String, dynamic> ? decoded : null;
  }

  static UdpDiscoveryAnnouncement? parseUdpDiscoveryMessage(String message) {
    final decoded = json.decode(message);
    if (decoded is! Map<String, dynamic>) {
      return null;
    }

    if (decoded['type'] != udpDiscoveryType) {
      return null;
    }

    final ip = decoded['ip'];
    final port = decoded['port'];
    final name = decoded['name'];
    if (ip is! String || port is! int || name is! String) {
      return null;
    }

    return UdpDiscoveryAnnouncement(ip: ip, port: port, name: name);
  }
}

class UdpDiscoveryAnnouncement {
  const UdpDiscoveryAnnouncement({
    required this.ip,
    required this.port,
    required this.name,
  });

  final String ip;
  final int port;
  final String name;
}
